earthfile2llb/converter.go

```diff
@@ -477,14 +476,25 @@ func (c *Converter) SaveImage(ctx context.Context, imageNames []string, pushImag
 		justCacheHint = true
 	}
 	for _, imageName := range imageNames {
+		var st llb.State
+		if pushImages && c.mts.Final.RunPush.Initialized() {
+			// SAVE IMAGE --push when it comes before any RUN --push should be treated as if they are in the main state,
+			// since thats their only dependency. It will still be marked as a push.
+			st = c.mts.Final.RunPush.State
+		} else {
+			st = c.mts.Final.MainState
+		}
+
 		c.mts.Final.SaveImages = append(c.mts.Final.SaveImages, states.SaveImage{
-			State:        c.mts.Final.MainState,
-			Image:        c.mts.Final.MainImage.Clone(),
-			DockerTag:    imageName,
-			Push:         pushImages,
-			InsecurePush: insecurePush,
-			CacheHint:    cacheHint,
+			State:               st,
+			Image:               c.mts.Final.MainImage.Clone(),
```

Comment:
Exactly right, and exactly why I did it this way.

I didn't want to also add an Image to RunPush as well, at that point we should probably go the extra step and make it a plain old SingleTarget since its adopting most fields...

Reply:
So do you want me to make that refactor?

Reply:
Either refactor or leave a note for our future selves?

Reply:
Left a note